src/mpl/include/mpl_thread_argobots.h

```diff
@@ -52,9 +52,42 @@ typedef void (*MPL_thread_func_t) (void *data);
 void MPL_thread_create(MPL_thread_func_t func, void *data, MPL_thread_id_t * idp, int *errp);
 
 #define MPL_thread_exit()
-#define MPL_thread_self(idp_) ABT_thread_self(idp_)
-#define MPL_thread_join(id_) ABT_thread_free(&id_)
-#define MPL_thread_same(idp1_, idp2_, same_)  ABT_thread_equal(*idp1_, *idp2_, same_)
+#define MPL_thread_self(idp_)                                                 \
+    do {                                                                      \
+        ABT_thread self_thread_tmp_ = ABT_THREAD_NULL;                        \
+        ABT_thread_self(&self_thread_tmp_);                                   \
+        uintptr_t id_tmp_;                                                    \
+        if (self_thread_tmp_ == ABT_THREAD_NULL) {                            \
+            /* It seems that an external thread calls this function. */       \
+            /* Use Pthreads ID instead. */                                    \
+            id_tmp_ = (uintptr_t)pthread_self();                              \
+            /* First, let's rotate bits since pthread_t might use the last    \
+             * bit (i.e., saving an ID starting from zero). */                \
+            id_tmp_ = (id_tmp_ << 1) |                                        \
+                      (id_tmp_ >> (sizeof(uintptr_t) * 8 - 1));               \
```

Comment:
You are going to overwrite the bit anyway, why bother to rotating it? The id returned by `MPL_thread_self` shold never be fed back to pthread functions so the comment `since pthread_t might use ...` is false.

Reply:
It is to avoid a conflict.

Reasonable `pthread_t` implementation can save an ID (i.e., not a pointer to a certain memory region) as `pthread_t`: I do not thoroughly check it, but Solaris uses 4-byte `pthread_t` on 64-bit machine, so it looks something other than a mere pointer. Because two `pthread_t`-based `MPL_thread_id_t`s must be different, we want to set a bit that is less likely to be used. Such a bit in `pthread_t` should be the first bit, not the last bit. That's why I rotate bits. If you think I am thinking too much, I can remove this rotation code.

Reply:
I can see the reason you want to "shift" to preserve the potential significant bits, I don't see why would you "or" the highest bit back since you are immediately overwriting it anyway.

Reply:
Oh I see. Thanks. Yes that's a good point I missed it. I just needed to shift it.

I will fix it.

Reply:
I fixed it. I removed the unnecessary ` (id_tmp_ >> (sizeof(uintptr_t) * 8 - 1));` part.